### ISSUE TITLE
$injector hook for prototypical inheritance

### ISSUE THREAD

**Issue by p-hebert** [angular/angular.js#14895]
It is currently nigh impossible/impossible to construct a classic prototypical chain with AngularJS constrained components (services, controllers, etc) due to complex restrictions on getting the access to the injection arguments for a given component.

By classic prototypical chain, I mean something akin the following:

```javascript
function Parent(){
  ...
}

function Child(){
 Parent.apply(this); 
 ...
}

Child.prototype = Object.create(Parent.prototype);

Child.prototype.constructor = Parent
```

With the current setup, in order to provide classic prototypical inheritance, one must create an injector, inject all necessary modules, get all individual injection dependencies for the constructor function of interest, and then proceed to create the prototype chain. This looks like the following (simplified, stripped of IIFEs for readability):

**Parent Controller**

```javascript
angular.module('app.parent', [/*dependencies*/]);
angular.module('app.parent').controller('Parent', Parent);

Parent.$inject = [/*some injectable dependencies*/];

function Parent(/*some injectable dependencies*/){
  ...
}
```

**Child Controller**

```javascript
angular.module('app.child', [/*dependencies*/]);
angular.module('app.child').controller('Child', Child);

$injector = angular.injector([ /* necessary modules to have all dependencies */]);
$super = $injector.get('Parent');
$dependencies = [];
$super.$inject.forEach(function(d) { $dependencies.push($injector.get(d)); });

function Child(/*some injectable dependencies*/){
 $super.apply(this, $dependencies);
 ... 
}

Child.prototype = Object.create($super.prototype);

Child.prototype.constructor = $super
```

I am fully aware that it is possible to provide *some level of inheritance* by using `angular.extend(this, $controller('Parent'))` in the Child constructor, but as it turns out this creates two objects and does not apply the logic of the Parent constructor on the Child. This can easily lead to some unexpected behaviour.

Unfortunately the solution proposed above does have drawbacks, and I have found myself unable to make it work *consistently*, due to dependencies declaration order problems (this can be resolved using something like RequireJS, but I'd rather avoid needing to integrate another framework for something as simple as prototype inheritance).

Hence I would like to propose an enhancement to the core of the AngularJS framework: the injector.
I am currently working with Angular 1.5.5, and the injector provides the following API:

```
      invoke: invoke,
      instantiate: instantiate,
      get: getService,
      annotate: createInjector.$$annotate,
      has: function(name) {
        return providerCache.hasOwnProperty(name + providerSuffix) || cache.hasOwnProperty(name);
      }
```

My proposition is to modify the injector's instantiate function signature in order to add a *thisArg* parameter. so that one can specify an argument on which to call the instantiation of the constructor passed to the injector:

```javascript
function instantiate(Type, locals, serviceName) {
      // Check if Type is annotated and use just the given function at n-1 as parameter
      // e.g. someModule.factory('greeter', ['$window', function(renamed$window) {}]);
      var ctor = (isArray(Type) ? Type[Type.length - 1] : Type);
      var args = injectionArgs(Type, locals, serviceName);
      // Empty object at position 0 is ignored for invocation with `new`, but required.
      args.unshift(null);
      return new (Function.prototype.bind.apply(ctor, args))();
}
```

would become

```javascript
function instantiate(Type, locals, serviceName, thisArg) {
      // Check if Type is annotated and use just the given function at n-1 as parameter
      // e.g. someModule.factory('greeter', ['$window', function(renamed$window) {}]);
      var ctor = (isArray(Type) ? Type[Type.length - 1] : Type);
      var args = injectionArgs(Type, locals, serviceName);
      // Empty object at position 0 is ignored for invocation with `new`, but required.
      args.unshift(typeof thisArg === "object" ? thisArg : null);
      return new (Function.prototype.bind.apply(ctor, args))();
}
```

and by extension, the `$controllerProvider.$get` `$controller` function could also receive a thisArg parameter, which could then be passed to the `$injector.instantiate` call to produce the same effect, all while retaining all core functionalities of AngularJS and not introducing any potential regression.

The reason I have not submitted this as a pull request is because I am curious if anyone actually already found an alternative solution to the problem. I am also curious if the ommission of such a functionality is a design choice from the AngularJS designers.

Let me know what you think!
Cheers

**Comment by gkalpak** [comment]
TBH, I don't think that inversion-of-control containers work well with this way of implementing inheritance.

If you don't want/can't use "standard" prototypal inheritance, I think you should manually pass the necessary arguments to `Parent`.

**Comment by lgalfaso** [comment]
Hi, I do not want to state the obvious, but did you notice the comment. `// Empty object at position 0 is ignored for invocation with new, but required.` ?

**Comment by gkalpak** [comment]
So, if I understand correctly, you basically don't need `$injector.instantiate` anyway. You need `$injector.invoke` (which already supports a `self` argument): [Demo][1]
(Thanks, @lgalfaso, for stating the obvious :smile:)

Does that solve your problem, @p-hebert?

[1]: https://codepen.io/ExpertSystem/pen/OXOEzv?editors=001

**Comment by p-hebert** [comment]
@lgalfaso: I wasn't sure of the meaning of the comment you pointed out - I inferred that `null` was pushed as the `this` argument of the Function.bind.apply function.

@gkalpak: I attempted something similar. I must concede that I didn't consider injecting the injector per say, but one of the major problems regarding this method is that it either requires the Parent constructor to be within the same IIFE/scope or to create an external $injector via `angular.injector` to load the `Parent` component, which then requires to load manually the module dependencies. The latter can become quickly problematic because it means that as soon as the IIFE is executed, the statement `$injector = angular.injector([ /* necessary modules to have all dependencies */]);` can produce a race condition with regards to module loading order. (In all honesty I have very little understanding of how to avoid such a race condition).

Let me know if I wasn't clear with my explanation!
Cheers

**Comment by gkalpak** [comment]
@p-hebert, I am still unsure how you would use the proposed feature (without having both constructors in the same IIFE. Maybe it would be simpler if you posted the code as you would ideally use it (assuming that the requested feature was already implemented).

**Comment by lgalfaso** [comment]
> @lgalfaso: I wasn't sure of the meaning of the comment you pointed out - I inferred that null was pushed as the this argument of the Function.bind.apply function.

TL; DR
The `null` value is ignored, as calling `new` only takes into consideration the type and the arguments, and not `this`.

Long version:
Calling `new` is equivalent to calling `EvaluateNew` with the type and the arguments
http://www.ecma-international.org/ecma-262/7.0/index.html#sec-evaluatenew

This calls `Construct` with the type and the arguments
http://www.ecma-international.org/ecma-262/7.0/index.html#sec-construct

This calls `[[Construct]]` with the arguments and `newTarget` and the type
http://www.ecma-international.org/ecma-262/7.0/index.html#sec-ecmascript-function-objects-construct-argumentslist-newtarget

`[[Construct]]` creates `thisArgument` based on `newTarget` and calls the function body with the arguments.

**Comment by p-hebert** [comment]
Alright,

so after reviewing @gkalpak suggestion and doing an [MVCE](http://codepen.io/p-hebert/pen/zBPVqo?editors=1111), it turns out that the current features implemented in Angular does support prototypical inheritance like what I am looking for - I will have to look more in depth regarding what makes my application crash with an error `Unknown provider: $compileProvider` as soon as I have an inheritance tree of height 2 and more.
